src/inference/gibbs.jl

```diff
@@ -2,7 +2,16 @@
 ### Gibbs samplers / compositional samplers.
 ###
 
-const GibbsComponent = Union{Hamiltonian,MH,ESS,PG}
+
+"""
+    isgibbscomponent(spl::A)
+    isgibbscomponent(::Type{A})
+
+Marks the type `S` to be allowed withing a `Gibbs` sampler.
+"""
+isgibbscomponent(::Type{<:InferenceAlgorithm}) = false
+isgibbscomponent(alg::InferenceAlgorithm) = isgibbscomponent(typeof(alg))
```

Comment:
It is recommended to define methods only for either instances or types, and recommended to define them for instances if possible (see https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-confusion-about-whether-something-is-an-instance-or-a-type-1). I would suggest to define only
```suggestion
isgibbscomponent(alg::InferenceAlgorithm) = alg
```
since it seems currently it is sufficient to be able to reason about instances.

Reply:
I know that phrase in the docs well, but every single time I write a function like this I think "oh, in this case it would surely be nice to make an exception" :P

Reply:
Another thing I was just wondering about: restricting to `InferenceAlgorithm` makes a narrower choice than the old `GibbsComponent`.  Is is even a good idea?

Reply:
I'm assuming it might not be, since there was this discussion that `GibbsConditional` should not be an `InferenceAlgorithm`, for example.

Reply:
Before it was fixed to `PG`, `Hamiltonian`, `MH`, and `ESS`, so it's not actually narrower, is it? And everyone could just implement `isgibbscomponent` for her custom algorithm, so it's not even generally restricted to `InferenceAlgorithm`. Whereas before it was fixed to the predefined algorithms.

Reply:
It was restricted to the end user, so to say, but it wasn't if you actually extended it internally.  The `GibbsConditional` in my other PR is not an `InferenceAlgorithm`, but still was added to `GibbsComponent`.  

The reason I'd at least add a non-restricted fallback is: if you have a new component that is not `<: InferenceAlgorithm`, and don't add the trait, you get a `MethodError`, instead of the error message from the constructor.  (Just happened to me :))

Reply:
> The reason I'd at least add a non-restricted fallback is: if you have a new component that is not <: InferenceAlgorithm, and don't add the trait, you get a MethodError, instead of the error message from the constructor. (Just happened to me :))

Yep, that's a good reason, I agree :+1:

> (the only way you could implement an algorithm was by changin Turing, anyway)

Is that true? Ideally you should be able to just define a new subtype of `InferenceAlgorithm`, implement the `AbstractMCMC` interface, and get access to all predefined dispatches (for `sample` etc).

I assumed the only/main problem with such a custom implementation would be that you can't hook into the Gibbs sampler (currently) due to `GibbsComponent` being defined internally.

Reply:
> I know that phrase in the docs well, but every single time I write a function like this I think "oh, in this case it would surely be nice to make an exception" :P

BTW same here :smile: I'm always inclined to implement both but then try to stick to the recommendations.

Reply:
> I assumed the only/main problem with such a custom implementation would be that you can't hook into the Gibbs sampler (currently) due to `GibbsComponent` being defined internally.

Yeah, that's what I meant.  As soon as you went so far as to change `GibbsComponent`, which you had to, you could put in any type.  But you're right, now that it works by dispatch it's open anyway.